Mathlib/Data/List/EditDistance/Estimator.lean

```diff
@@ -0,0 +1,143 @@
+provides successively better lower bounds for the Levenshtein distance as it runs,
+as proved in `Mathlib.Data.List.EditDistance.Bounds`.
+
+In this file we package that fact as an instance of
+```
+Estimator (Thunk.mk fun _ => levenshtein C xs ys) (LevenshteinEstimator C xs ys)
+```
+allowing us to use the `Estimator` framework for Levenshtein distances.
+
+This is then used in the implementation of `rewrite_search`
+to avoid needing the entire edit distance calculation in unlikely search paths.
+-/
+
+set_option autoImplicit true
+
+variable [CanonicallyLinearOrderedAddMonoid δ]
+    (C : Levenshtein.Cost α β δ) (xs : List α) (ys : List β)
+
+/--
+Data showing that the Levenshtein distance from `xs` to `ys`
+is bounded below by the minimum Levenshtein distance between some suffix of `xs`
+and a particular suffix of `ys`.
+
+This bound is (non-strict) monotone as we take longer suffixes of `ys`.
+
+This is an auxiliary definition for the later `LevenshteinEstimator`:
+this variant constructs a lower bound for the pair consisting of
+the Levenshtein distance from `xs` to `ys`,
+along with the length of `ys`.
+-/
+structure LevenshteinEstimator' : Type _ where
+  /-- The prefix of `ys` that is not is not involved in the bound, in reverse order. -/
+  pre_rev : List β
+  /-- The suffix of `ys`, such that the distance from `xs` to `ys` is bounded below
+  by the minimum distance from any suffix of `xs` to this suffix. -/
+  suff : List β
+  /-- Witness that `ys` has been decomposed into a prefix and suffix. -/
+  split : pre_rev.reverse ++ suff = ys
+  /-- The distances from each suffix of `xs` to `suff`. -/
+  distances : {r : List δ // 0 < r.length}
+  /-- Witness that `distances` are correct. -/
+  distances_eq : distances = suffixLevenshtein C xs suff
+  /-- The current bound on the pair (distance from `xs` to `ys`, length of `ys`). -/
+  bound : δ × ℕ
+  /-- Predicate describing the current bound. -/
+  bound_eq : bound = match pre_rev with
+    | [] => (distances.1[0]'(distances.2), ys.length)
+    | _ => (List.minimum_of_length_pos distances.2, suff.length)
+
+instance : EstimatorData (Thunk.mk fun _ => (levenshtein C xs ys, ys.length))
+    (LevenshteinEstimator' C xs ys) where
+  bound e := e.bound
+  improve e := match e.pre_rev, e.split with
+    | [], _ => none
+    | y :: ys, split => some
+      { pre_rev := ys
+        suff := y :: e.suff
+        split := by simpa using split
+        distances := Levenshtein.impl C xs y e.distances
+        distances_eq := e.distances_eq ▸ suffixLevenshtein_eq xs y e.suff
+        bound := _
+        bound_eq := rfl }
+
+instance estimator' :
+    Estimator (Thunk.mk fun _ => (levenshtein C xs ys, ys.length))
+      (LevenshteinEstimator' C xs ys) where
+  bound_le e := match e.pre_rev, e.split, e.bound_eq with
+  | [], split, eq => by
+    simp only [List.reverse_nil, List.nil_append] at split
+    rw [e.distances_eq] at eq
+    simp only [List.getElem_eq_get] at eq
+    rw [split] at eq
```

Comment:
What's happening here? Can't you reduce this to a single `simp only` or at least a `simp_rw`?

Reply:
Unfortunately not: changing the first `rw` to a `simp_rw` causes the last `rw` to error with a "motive not type correct". `simp` doesn't make any progress on the when used in place of the last `rw`.

Ugly, hey?